Add selectrum quickselect keys.

Luis-Henriquez-1 opened the pull request:
<!--

To expedite the pull request process, please see the contributor guide
for my projects:

  <https://github.com/raxod502/contributor-guide>

-->

This purpose of this pull request is to make selectrum display indices only on visible candiates and to provide the user with a way to customize the character used for the indices of candidates.

To do this I created (as suggested by @okamsn in #194 ) an alist called `selectrum-quick-selet-keys` whose keys are indices and whose values are the characters that should be displayed for that index.

The name of the variable `selectrum-show-indices` was no longer well suited to its function, therefore I renamed it to  `selectrum-show-quick-select-keys`.

There are a few open questions however.

One is, should the alist use strings as values instead of characters?

Using characters works file for single letters. However, if the user prefers using numbers and has `selectrum-num-candidates-displayed` greater than 10, then there will be double digits and therefore it might be less of a pain using a string.

Another is, should we get rid of the variable `selectrum-show-quick-select-keys` all-together?

Technically we only need the alist. We can just set it to nil by default.

I'm sure there are more things I haven't thought of. So I expect to have to make many more changes.

Finally, I want to make sure to give credit to @okamsn  and @clemera  who were very responsive and helpful when I created this issue.

clemera commented:
Thanks! I think we can keep the `selectrum-show-indices` option and configure this feature using an additional `selectrum-count-style`? I think for now we should focus on implementing a new display style (maybe called `relative`) which annotates displayed rows by their relative position. Later we can think about how to additionally let users configure custom chars or allow to have direct bindings to choose those.

clemera commented:
I would still favour to keep `selectrum-show-indices` so we don't break user configuration. Maybe we could simply also allow a list of chars as its value? I don't think we need an alist because the keys are simply the index positions.

Luis-Henriquez-1 commented:
> I would still favour to keep selectrum-show-indices so we don't break user configuration.
> I don't think we need an alist because the keys are simply the index positions.

Ok, I'm completely fine with that. 

> Maybe we could also allow a list of chars as its value?

With the list of chars, what happens when if the user wants some two character value?

Suppose the user had `selectrum-num-candidates-displayed` as 15 and wanted numbered candidates, then a character wouldn't be able to represent the "15". Should we have strings instead?

okamsn commented:
I think it makes sense to top-down a bit, and decide the top-level details before trying to figure out what the list should contain.

For example, I think it makes sense to have a separate variable to determine whether the displayed prefixes should change when scrolling, and that this should apply to the numbered indices just as it would for lettered indices. If done, then I think it makes sense to adjust the `selectrum-select-current-candidate` and `selectrum-insert-current-candidate` commands to handle that case when given numeric arguments. This way, one can give the normal `C-u 3 ...` without this pull request needing to worry about using strings or characters in the list. The features for handling numeric indices already exist, and it looks like Selectrum can tell which candidate is displayed first according to the line `(let* ((first-index-displayed` near 784. I think that is sufficient for the case of numeric indices.

For letter prefixes, do you expect people will want to use, for example, multiple easily reached letters like `"aa"` more than letters that are father away from the home row, like `?p` or `?q`.  `read-char` is easy to use, and means needing only two key presses to select a candidate. There might be a better way than the example I gave, but it was the first method to occur to me.

There's also the case of using letters but allowing the indices to move with the candidates. Should the candidates just be numbered after running out of letters, or maybe using a function is better than a list in this case?

What do you think about these things?

clemera commented:
> For example, I think it makes sense to have a separate variable to determine whether the displayed prefixes should change when scrolling, and that this should apply to the numbered indices just as it would for lettered indices. If done, then I think it makes sense to adjust the selectrum-select-current-candidate and selectrum-insert-current-candidate commands to handle that case when given numeric arguments. This way, one can give the normal C-u 3 ... without this pull request needing to worry about using strings or characters in the list when dealing with numbers. The features for handling numeric indices already exist, and it looks like Selectrum can tell which candidate is displayed first according to the line (let* ((first-index-displayed near 784. I think that is sufficient for the case of numeric indices.


I just tested and it seems selectrum-select-current-candidate and selectrum-insert-current-candidate commands already handle numeric args correctly so I don't understand why they would need to be adjusted.

I'm not sure if it makes sense to have non relative indices at all. Is there a reason one would like to see the exact index of a candidate? If not it might make sense to change the current default so the indices always show the current row/numeric argument. There could be an additional option to specify custom chars which defaults to nil.

> Suppose the user had selectrum-num-candidates-displayed as 15 and wanted numbered candidates, then a character wouldn't be able to represent the "15". Should we have strings instead?

I think the purpose of a custom char list would be quick access and you probably wouldn't like to specify two chars? Thinking about it I'm not sure we need an option to specify custom chars aren't numeric args sufficient? The users which specify custom chars would additionally need to bind commands to choose indices via their custom char mappings.

Luis-Henriquez-1 commented:
> I'm not sure if it makes sense to have non relative indices at all.

This is what I was going to say. I also can't think of any advantage of non-relative indices.

> Thinking about it I'm not sure we need an option to specify custom chars, aren't numeric args sufficient?

I know I personally prefer setting the characters to letters so that I can use the homerow keys. When I used ivy, I'd use [ivy-avy](https://github.com/abo-abo/swiper/blob/master/ivy-avy.el) all the time to select candidates.

> The users which specify custom chars would additionally need to bind commands to choose indices using their custom char mappings.

That's true. I wouldn't mind having to do this.

> I think the purpose of a custom char list would be quick access and you probably wouldn't like to specify two chars?

True. Perhaps my thinking is mistaken. I was under the impression we'd just use this alist/function for everything. The idea being we wouldn't have both a default case and a generic way of specifying the indices--the default is just a special case of the generic way.

For example, this is what I envisioned.

We'd add one variable `selectrum-show-indices-fn` by default it would be set to `(lambda (index) (format "%d" index))`. When `selectrum-show-indices` is t, we use this function to display indices. When it's nil, we don't.

```elisp
;; as a function
(defcustom selectrum-show-indices-fn
  (lambda (index) (format "%d" index)))
```

> or maybe using a function is better than a list in this case?

I personally favor using a function. I think it would be the most concise way of specifying a pattern. It can take in the index of a displayed candidate and return the string/character/number that should be displayed to the left of the candidate.

clemera commented:
I think using a function is nice. We can also avoid adding a new option by allowing a function value for `selectrum-show-indices`. We can pass the absolute index and the total number of candidates as arguments to the function, too so they can be easily used if needed. I think we can switch to show relative indices when the option is `t` because that seems like a better default. To restore the old behaviour we can provide a function which we mention in the docstring.

Luis-Henriquez-1 commented:
@clemera I've implemented some of the ideas I suggested to see how they would work in practice. I did it before I read your post just now, but I can easily change things to accommodate your suggestions.

This pull request can be divided into two major topics.

A) showing the indices of only candidates displayed.

B) providing a customization for how the displayed indices look

In my changes I added one commit which only deals with A. 

The other two commits deal with B. 

The first of the two  just implements adding `selectrum-show-indices-fn`.

The second of the two I realized that it makes more sense to include padding in the show indices function, because we don't know how the user wants the stuff spaced.

I showcase the results in this giff.

![selectrum-indice-customization](https://user-images.githubusercontent.com/30783193/92958322-05bf7200-f438-11ea-8243-329eb2c026b2.gif)

> we can pass the absolute index and the total number of candidates

Ok, I used just the index, but I'll add the total number of candidates.

> we can avoid adding a new option

Ideally, it would be best to just have one variable `selectrum-show-indices-fn` that can return the empty string if no indices should be visible. _However,_ given that we highly value not breaking existing configurations, I favor having the two variables over adding more possible values to `selectrum-show-indices`. My reasoning is that `selectrum-show-indices` starts to lose its original meaning which was just a boolean on/off switch. Since we're answering two questions: whether indices should be displayed and how they should be displayed it seems that we need two variables. 

That being said, I don't think this is a huge deal and I will gladly do this if you think it is best.

> To restore the old behavior

Ok, I don't think I provided a way to restore the old behavior in my commits as I belive the function takes in the relative index. I'll change this.

clemera commented:
> Ideally, it would be best to just have one variable selectrum-show-indices-fn that can return the empty string if no indices should be visible. However, given that we highly value not breaking existing configurations, I favor having the two variables over adding more possible values to selectrum-show-indices. My reasoning is that selectrum-show-indices starts to lose its original meaning which was just a boolean on/off switch. Since we're answering two questions: whether indices should be displayed and how they should be displayed it seems that we need two variables.

I would prefer that users have a single place to configure this. I think it's better to keep the number of options small and I don't see a benefit introducing another option for this.

> I don't think I provided a way to restore the old behavior in my commits as I belive the function takes in the relative index. I'll change this.

I'm not sure it's needed, as I said I don't know the benefits of displaying the absolute index. @raxod502 added it maybe he has a use case for this?

okamsn commented:
> > For example, I think it makes sense to have a separate variable to determine whether the displayed prefixes should change when scrolling, and that this should apply to the numbered indices just as it would for lettered indices. If done, then I think it makes sense to adjust the selectrum-select-current-candidate and selectrum-insert-current-candidate commands to handle that case when given numeric arguments. This way, one can give the normal C-u 3 ... without this pull request needing to worry about using strings or characters in the list when dealing with numbers. The features for handling numeric indices already exist, and it looks like Selectrum can tell which candidate is displayed first according to the line (let* ((first-index-displayed near 784. I think that is sufficient for the case of numeric indices.
> 
> I just tested and it seems selectrum-select-current-candidate and selectrum-insert-current-candidate commands already handle numeric args correctly, why do you think they need to be adjusted?

I meant adjusting for the case of relative indices, sine they currently seem to work using the absolute index (plus 1). I agree though that just using a relative index by default feels more useful.

If you allow for using letters as indices, what do you think is a good way to turn key presses back into the needed index?

Luis-Henriquez-1 commented:
> I meant adjusting for the case of relative indices, sine they currently seem to work using the absolute index (plus 1)

Is the relative index the absolute index plus one? I'm confused. Could you post your code for this?

> If you allow for using letters as indices, what do you think is a good way to turn key presses back into the needed index?

Do you mean turning them back for binding them to keys? Or for converting from relative to absolute (or vice versa)? If the latter, when would we need to convert it?

Also, based on the issue I authored #202, and some I read (#181, #156) I'm wondering we need this index stuff. If a framework is made to display candidate metadata I don't see why this shouldn't display the index as well. Clemera is more knowledgeable about this. @clemera, Are the changes that are in development concerning displaying icons going to allow selectrum to display arbitrary metadata (such as the image of ivy I posted in #202)? And if so is this going to be responsible for displaying indices too?

okamsn commented:
I apologize for my confusing phrasing. I will try to be clearer.

> > I meant adjusting for the case of relative indices, sine they currently seem to work using the absolute index (plus 1)
> 
> Is the relative index the absolute index plus one? I'm confused. Could you explain or post the code for this?

No, by "they", I mean the commands `selectrum-select-current-candidate` and `selectrum-insert-current-candidate`. Their optional parameter is the absolute index displayed by Selectrum, which starts counting at one, so that the candidate listed as "1" is the zeroth item in the actual list. I was saying that if Selectrum were configured to use either absolute indices (as it does now) or relative indices (starting from the top of the minibuffer window), then I think that the commands should be adjusted to be able to work in both cases. However, if Selectrum is changed to only use the relative indices, then it is a moot point.

> EDIT: Maybe we should use the `%` operator so we could use absolute index but and still get the relative index if we wanted it.
> 
> > If you allow for using letters as indices, what do you think is a good way to turn key presses back into the needed index?
> 
> Do you mean turning them back for binding them to keys? Or for converting from relative to absolute (or vice versa)? If the latter, when would we need to convert it?

In a way, I meant both. I'm still thinking of this in terms of the example I gave, which used `read-key` and the alist. As I currently understand it, with relative indices (which I didn't implement in the example), you need a way to associate a key or letter with the relative index, and then a way to to associate that relative index with the actual candidate. You can see that `selectrum--get-candidate`, used by the above mentioned commands for insertion and selection, uses the normal, starting-at-zero kind of index.

clemera commented:
> Are the changes that are in development concerning displaying icons going to allow selectrum to display arbitrary metadata (such as the image of ivy I posted in #202)? And if so is this going to be responsible for displaying indices too?

AFAIK there is no concrete development of this feature for now. I don't think indices should be handled by such mechanism because they are a general feature for selecting a candidate which should be independent of the candidate display format. Otherwise when we would allow custom display functions people would have to handle indices display in all their custom display functions.

Luis-Henriquez-1 commented:
> I don't think indices or the selection of indices should be handled by such mechanism because they are a general feature for selecting a candidate

I'm not saying that a mechanism should handle the indices themselves or the selection of indices. I just
think it should handle the display of an index (if the user wants to display it).

> people would have to handle indices display in all their custom display functions

I would assume there would be a default display function used for commands that
they didn't specify a custom display. The display function could just be
something like `(lambda (i candidate) (format "%2d %s"))` if they want indices
by default.

Moreover, users should want the ability to handle indices display in their
custom display functions. I can easily imagine adventurous users wanting to
play (rice) with the design of the display which may very well include the location
of the index. Maybe having it at the end of the line, or the middle of the line.
Or maybe in different locations for different commands.

Another reason I think indices should be apart of it is because the alternative
is a duality. Now there are multiple things affecting the display of a command:
`selectrum-show-indices` and the (yet to be concretely defined) display
framework. Which means there are two places the user might need to customize to
achieve a certain display.

This being said, the most important thing is that such a mechanism would at
least have the capability of displaying the candidate index. Having a separate
way to display indices would then be unnecessary but benign considering the user
can just disable it and do it themselves using the display mechanism.

clemera commented:
> Moreover, users should want the ability to handle indices display in their
custom display functions. I can easily imagine adventurous users wanting to
play (rice) with the design of the display which may very well include the location
of the index. Maybe having it at the end of the line, or the middle of the line.
Or maybe in different locations for different commands.

I think we should avoid a situation where the display position depends on the implementation of a command. As a user you probably prefer a consistent interface than having to search where a command might have placed the index.

> Another reason I think indices should be apart of it is because the alternative
is a duality. Now there are multiple things affecting the display of a command:
selectrum-show-indices and the (yet to be concretely defined) display
framework. Which means there are two places the user might need to customize to
achieve a certain display.

I think of indices display as a general UI decision. Users should be able to decide if they want them to be displayed or not like we have it currently. If there are custom display functions for commands those display functions would need to check for this user option to decide if they should be included.

Luis-Henriquez-1 commented:
> I think we should avoid a situation where the display position depends on the command. 

I did not suggest that we should seek suck a situation. That was one of three hypothetical examples I provided as something that a user could do with the index. My point was that the display should be very customizable and that knowing the index can be very useful for the customization of the display.

> As a user you probably prefer a consistent interface than having to search where a command might place the index.

If a user chooses to always have their index at the right, that would also be a consistent interface. But right now selectrum always places it at the beginning of the line. Again, I just want to maximize user customization and reduce locking them down to any bias (like having an index on the left).

> If there are custom display functions which also handle index display these functions would need to check for the index display option to decide if the indices should be shown.

The way I envision it is that the functions would not need (and should not) check for this.

These functions wouldn't do anything to indices by default.

If a user desires to customize their display enough to be willing to read selectrum documentation on how to do it, surely he won't mind setting `selectrum-show-indices` to nil if we wants a custom index display.

Note that if you don't mention an index in the custom display function(s), then `selectrum-show-indices` would just prepend an index to the right of whatever these functions choose to display. 


In any case, perhaps I'm digressing too much from the issue at hand. I'll finish up your suggestion about getting rid of `selectrum-show-indice-fn` and update the readme today so we can get this merged. And we can move the discussion of the custom display into a potential issue for it (I may create) as it is something I would be interested in.

clemera commented:
> I did not suggest that we should seek suck a situation. That was one of three hypothetical examples I provided as something that a user could do with the index. My point was that the display should be very customizable and that knowing the index can be very useful for the customization of the display.

Okay, I was responding to the example which talked about command specific indices. 

> If a user chooses to always have their index at the right, that would also be a consistent interface. But right now selectrum always places it at the beginning of the line. Again, I just want to maximize user customization and reduce locking them down to any bias (like having an index on the left).

My consistency argument was still in the response to command specific indices. In general I don't think index display on other positions would be a common desire because you probably want them at a predictable position not to far from where your eyes are. 

> In any case, perhaps I'm digressing too much from the issue at hand. I'll finish up your suggestion about getting rid of selectrum-show-indice-fn and update the readme today so we can get this merged. And we can move the discussion of the custom display into a potential issue for it (I may create) as it is something I would be interested in.

Customization always increases complexity and I prefer to keep things simple for the most common use case. As I see it there are other issues with regard to highlighting which one would need to handle when having full freedom over index display but I'm open to discuss any code if you care to implement it. 

This all is of course only my opinion.

Luis-Henriquez-1 commented:
> Customization always increases complexity and I prefer to keep things simple for the most common use case. As I see it there are other issues with regard to highlighting which one would need to handle when having full freedom over index display but I'm open to discuss any code if you care to implement it.

> This all is of course only my opinion.

 Ok, I respect your opinion. Thank you for hearing me out and taking time to respond.

When I was updating the code for selectrum I ran into a problem.

1) Including the maximum number of candidates for the show indices function means that I'll have to get the length of the list of candidates in `selectrum--candidates-display-string'. I'm concerned about performance. I know this function is called many times, so adding a `(length N)` call every time will make selectrum slower. Even if it's not noticable right now, its these kind of things that add up.

Is adding the maximum number of candidates as a parameter designed to let the user know how much padding to add for the custom indices? I'm thinking if we are going to have relative indices anyway, maybe we could just have the index parameter.

```elisp
(let* ((index 0)
         (lines
          (selectrum--ensure-single-lines
           (funcall selectrum-highlight-candidates-function
                    input candidates)))
       (num-lines (length lines))
    ...
    )
```

This is what I've got so far for the `selectrum-show-indices` application. At first I wasn't sure whether `functionp` would be able to tell that a `lambda` form is a function but it seems to be able to do so.

```elisp
(when selectrum-show-indices
            (let* ((abs-index (+ index 0))
                   (display-fn (if (functionp selectrum-show-indices)
                                   selectrum-show-indices
                                 (lambda (i) (format "%2d " i))))
                   (curr-index (funcall display-fn (1+ abs-index))))
              (insert
               (propertize curr-index 'face 'minibuffer-prompt))))
```

clemera commented:
> Including the maximum number of candidates for the show indices function means that I'll have to get the length of the list of candidates in selectrum--candidates-display-string. I'm concerned about performance. I know this function is called many times, so adding a (length N) call every time will make selectrum slower. Even if it's not noticable right now, its these kind of things that add up.

> Is adding the maximum number of candidates as a parameter designed to let the user know how much padding to add for the custom indices? I'm thinking if we are going to have relative indices anyway, maybe we could just have the index parameter.

Previously the length was captured in `selectrum--total-num-candidates` on input change in post-command-hook. `length` is implemented in C so this shouldn't effect the user experience:

```elisp
(let ((cands (all-completions "" 'help--symbol-completion-table)))
  (benchmark 10000 (length cands)))
```

But I'm also not sure if we need to support the old behaviour because I don't know the use case of showing the absolute index. Users which have set this option will hopefully see the new behaviour as an improvement, if someone complaints we can still consider readding it. I think we can do as you suggest and just use the index parameter.

> This is what I've got so far for the selectrum-show-indices application. At first I wasn't sure whether functionp would be able to tell that a lambda form is a function but it seems to be able to do so.

Looks good to me, the `abs-index` variable can be removed because it's not needed any more. Maybe we should use `substring-no-properties` to make sure `display-fn` can't return a string with any other fancy display behaviour.

Luis-Henriquez-1 commented:
My recent commits contain the following changes:

1- adding relative indices by default
2- allowing selectrum-show-indices to be a function
3- updating the readme and changelog

I'm not the best at writing docstrings. Please read over what I've done when you get the time.

okamsn commented:
In the doc-strings, would you mind specifying that "relative" means from the top of the window? I feel like that would be more informative.

Luis-Henriquez-1 commented on README.md:
```diff
@@ -282,6 +282,10 @@ matching and case-insensitive matching.
+* If you want do display a custom index (e.g. letters instead of
+  indices, roman numerals, etc.) you can set the
+  `selectrum-show-indices` to a function that takes in the relative
+  index of a candidate and returns the string you want to display.
```
I amended my last commit so that it's added it to the previous list item.

Luis-Henriquez-1 commented on selectrum.el:
```diff
@@ -250,9 +250,12 @@ Possible values are:
+displayed candidate as a parameter and it should return the string to
+be displayed (presumably) representing the index of the candidate. If
+this is some other non-nil value, it is treated as if it were
+(lambda (i) (format \"%2d \" i))."
```
> first line position need to be escaped in Elisp

It's now no longer in the first line position. I don't know about the selectrum whitespace convention. I used `fill-paragraph` to space the docstring.

> relative index term might not be clear to users, maybe we could use something like the "row number of the displayed candidate"

I updated this to your suggestion.

Luis-Henriquez-1 commented:
Thank you @clemera and @okamsn  for reviewing my changes. I have amended my commit accordingly.

clemera commented:
Thank you! Could you fix what `make lint` reports?

Luis-Henriquez-1 commented:
`make lint` reports unused lexical argument `first-index-displayed` and it's referring to the argument used in `selectrum--candidates-display-string`.i

I get the following output from `make lint`.

```
  -> make lint
[compile] selectrum.el
[compile] selectrum-helm.el
[checkdoc] selectrum.el
[checkdoc] selectrum-helm.el
[longlines] CHANGELOG.md
[longlines] .circleci/config.yml
[longlines] .dir-locals.el
[longlines] Dockerfile
[longlines] .github/PULL_REQUEST_TEMPLATE.md
[longlines] .gitignore
[longlines] LICENSE.md
[longlines] Makefile
[longlines] README.md
[longlines] scripts/check-line-length.bash
[longlines] scripts/docker.bash
[longlines] scripts/docker-install.bash
[longlines] selectrum.el
[longlines] selectrum-helm.el
[checkindent] selectrum.el
[checkindent] selectrum-helm.el
[toc] README.md
  --> markdown-toc missing, skipping
```

Is that it or is there anything else that needs fixing?

clemera commented:
Thanks, this should be all. There is currently a problem with the master build. We should wait a few days if @raxod502 has something to add in regard to absolute vs relative indices.

okamsn commented:
What about the commands `selectrum-select-current-candidate` and `selectrum-insert-current-candidate`? Would you please change them to also work with the relative indices?

I haven't looked hard at it, but you can copy the code from `selectrum--minibuffer-post-command-hook` that uses the `selectrum--clamp` function to find the candidate displayed at the top of the window.

The following is what I did to make it work with relative indices (really just copy and paste), but I haven't checked that all of it is necessary.

```diff
@@ -1224,8 +1250,21 @@
 Zero means to select the current user input."
   (interactive "P")
   (let ((index (if arg
-                   (min (1- (prefix-numeric-value arg))
-                        (1- (length selectrum--refined-candidates)))
+                   (min
+                    (1- (+ (if selectrum--current-candidate-index
+                               (selectrum--clamp
+                                ;; Adding one here makes it look slightly better, as
+                                ;; there are guaranteed to be more candidates shown
+                                ;; below the selection than above.
+                                (1+ (- selectrum--current-candidate-index
+                                       (max 1 (/ selectrum-num-candidates-displayed 2))))
+                                0
+                                (max (- (length selectrum--refined-candidates)
+                                        selectrum-num-candidates-displayed)
+                                     0))
+                             0)
+                           (prefix-numeric-value arg)))
+                    (1- (length selectrum--refined-candidates)))
                  selectrum--current-candidate-index)))
     (when (or (not selectrum--match-required-p)
               (and index (>= index 0))
@@ -1270,8 +1309,21 @@
   (if-let ((index (if (and arg
                            selectrum--refined-candidates
                            (> (prefix-numeric-value arg) 0))
-                      (min (1- (prefix-numeric-value arg))
-                           (1- (length selectrum--refined-candidates)))
+                      (min
+                       (1- (+ (if selectrum--current-candidate-index
+                                  (selectrum--clamp
+                                   ;; Adding one here makes it look slightly better, as
+                                   ;; there are guaranteed to be more candidates shown
+                                   ;; below the selection than above.
+                                   (1+ (- selectrum--current-candidate-index
+                                          (max 1 (/ selectrum-num-candidates-displayed 2))))
+                                   0
+                                   (max (- (length selectrum--refined-candidates)
+                                           selectrum-num-candidates-displayed)
+                                        0))
+                                0)
+                              (prefix-numeric-value arg)))
+                       (1- (length selectrum--refined-candidates)))
                     selectrum--current-candidate-index))
            (candidate (nth index
                            selectrum--refined-candidates))
```

raxod502 commented:
What a great discussion and collaboration! It is really beautiful to see such vibrant work happening even when I do not have the time to drop by. I do not think I actually have much to add to where the thread is right now, other than I think it's totally fine to switch from absolute to relative indices. While I won't claim that nobody will complain...

<a href="https://xkcd.com/1172/"><img src="https://imgs.xkcd.com/comics/workflow_2x.png" width="300px"></a>

... I think it is a very sensible improvement.

clemera commented:
Thanks @okamsn! I think we can simplify the index computation to:

```elisp
(if arg
    (min
     (+ (prefix-numeric-value arg)
        selectrum--current-candidate-index)
     (1- (length selectrum--refined-candidates)))
  selectrum--current-candidate-index)
```

Luis-Henriquez-1 commented:
> What about the commands selectrum-select-current-candidate and selectrum-insert-current-candidate? Would you please change them to also work with the relative indices?

I just pushed the changes recommended by @okamsn and simplified by @clemera.

raxod502 commented:
If you rebase onto the latest master then it should fix the build failure for emacs-master.

clemera commented:
Sorry for the delay, in final review I noticed the `:type` of the `defcustom` should be changed to `(choice function boolean)`. Lastly the changelog entries currently don't reference this PR. With these fixed I think this is ready to merge.

Luis-Henriquez-1 commented:
> With these fixed I think this is ready to merge.

Ok. Fixed them.

clemera commented:
Have you noticed the failing tests?

Luis-Henriquez-1 commented:
Ok. One line was too long. All checks are passing now.

clemera commented:
Thank you!
